### DISCUSSION TITLE
How to call a function after update state?

### DISCUSSION THREAD

**Discussion by Toshinaki** [pmndrs/zustand#1311] [Q&A] [answered]
I'm experimenting zustand and want to send an http request after update the store locally.
(I'm using `immer` middleware)

```js
{
  user: {id, name, ...},
  updateUser: (data) => set((state) => {
    state.user = data
  })
}
```

After calling `updateUser`, I want to send the updates to server.

Can I do (1)
```js
updateUser: (data) => set((state) => {
  state.user = data;
  updateServerUser(data);
})
```

Or should I define a hook (2):
```js
const useUpdateUser = () => {
  const {updateUser} = useStore();
  return (data) => {
    updateUser(data);
    updateServer(data);
  }
}
```

Or just a plain function (3):
```js
const updateUser = (data) => {
  useStore().updateUser(data);
  updateServer(data);
}
```

What's the best solution?

Or is there any better one?

**Comment by dai-shi** [comment, marked as answer]
They all work. I would prefer (1) or (3) depending on the actions style: https://github.com/pmndrs/zustand/blob/main/docs/guides/practice-with-no-store-actions.md

creating a hook like (2) is probably unnecessary.

**Reply by Toshinaki** [reply]
Nice!
Splitting state and actions is more reasonable to me.

BTW, this example from the link:
```js
export const useBoundStore = create(() => ({
  count: 0,
  text: 'hello',
}))

export const inc = () =>
  useBoundStore.setState((state) => ({ count: state.count + 1 }))

export const setText = (text) => useBoundStore.setState({ text })
```
The default state is `{count: 0, text: 'hello'}`, after calling `inc`, the state will be `{count: 1, text: 'hello'}`, not `{count: 1}`, right?

**Reply by dai-shi** [reply]
Yes, it merges states by default.

**Reply by Toshinaki** [reply]
Many thanks!! Loving this package.
